File: python/071-080/Edit Distance.py
class Solution:
    # @param {string} word1
    # @param {string} word2
    # @return {integer}
    def minDistance(self, word1, word2):
        n,m = len(word1), len(word2)
        p = [[0]*(m+1) for _ in range(n+1)]
        
        for i in range(1, n+1):
            p[i][0] = i
        for i in range(1, m+1):
            p[0][i] = i
        
        for i in range(1, n+1):
            for j in range(1, m+1):
                f = 1 if word1[i-1] != word2[j-1] else 0
                p[i][j] = min(p[i-1][j]+1, p[i][j-1]+1, p[i-1][j-1]+f);
        
        return p[n][m]

File: cpp/071-080/Edit Distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.length();
        int n = word2.length();
        
        vector<vector<int>> p(m+1, vector<int>(n+1));

        for (int i = 0;i <= m;i++)
            p[i][0] = i;
        for (int j = 0;j <= n;j++)
            p[0][j] = j;
        
        for (int i = 1;i <= m;i++) {
            for (int j = 1;j <= n;j++) {
                
                if (word1[i-1] == word2[j-1]) {
                    p[i][j] = p[i-1][j-1];
                } else {
                    p[i][j] = min(1 + p[i-1][j-1], 1+min(p[i][j-1], p[i-1][j]));
                }
                
            }
        }
        
        return p[m][n];
    }
};
